pkg/k8s/watchers/cilium_cidr_group.go

```diff
@@ -0,0 +1,253 @@
+// SPDX-License-Identifier: Apache-2.0
+// Copyright Authors of Cilium
+
+package watchers
+
+import (
+	"errors"
+	"fmt"
+	"time"
+
+	"github.com/sirupsen/logrus"
+
+	cilium_v2_alpha1 "github.com/cilium/cilium/pkg/k8s/apis/cilium.io/v2alpha1"
+	"github.com/cilium/cilium/pkg/k8s/client"
+	"github.com/cilium/cilium/pkg/k8s/resource"
+	"github.com/cilium/cilium/pkg/k8s/types"
+	"github.com/cilium/cilium/pkg/k8s/watchers/resources"
+	"github.com/cilium/cilium/pkg/logging/logfields"
+	"github.com/cilium/cilium/pkg/metrics"
+	"github.com/cilium/cilium/pkg/policy/api"
+)
+
+func (k *K8sWatcher) onUpsertCIDRGroup(
+	cidrGroup *cilium_v2_alpha1.CiliumCIDRGroup,
+	cidrGroupCache map[string]*cilium_v2_alpha1.CiliumCIDRGroup,
+	cnpCache map[resource.Key]*types.SlimCNP,
+	cs client.Clientset,
+	apiGroup, metricLabel string,
+) error {
+	var (
+		equal  bool
+		action string
+	)
+
+	// wrap k.K8sEventReceived call into a naked func() to capture equal in the closure
+	defer func() {
+		k.K8sEventReceived(apiGroup, metricLabel, action, true, equal)
+	}()
+
+	oldCidrGroup, ok := cidrGroupCache[cidrGroup.Name]
+	if ok && oldCidrGroup.Spec.DeepEqual(&cidrGroup.Spec) {
+		equal = true
+		return nil
+	} else if ok {
+		action = resources.MetricUpdate
+	} else {
+		action = resources.MetricCreate
+	}
+
+	cidrGroupCpy := cidrGroup.DeepCopy()
+	cidrGroupCache[cidrGroup.Name] = cidrGroupCpy
+
+	err := k.updateCIDRGroupRefPolicies(cidrGroup.Name, cidrGroupCache, cnpCache, cs)
+
+	k.K8sEventProcessed(metricLabel, action, err == nil)
+
+	return err
+}
+
+func (k *K8sWatcher) onDeleteCIDRGroup(
+	cidrGroupName string,
+	cidrGroupCache map[string]*cilium_v2_alpha1.CiliumCIDRGroup,
+	cnpCache map[resource.Key]*types.SlimCNP,
+	cs client.Clientset,
+	apiGroup, metricLabel string,
+) error {
+	delete(cidrGroupCache, cidrGroupName)
+
+	err := k.updateCIDRGroupRefPolicies(cidrGroupName, cidrGroupCache, cnpCache, cs)
+
+	k.K8sEventProcessed(metricLabel, resources.MetricDelete, err == nil)
+	k.K8sEventReceived(apiGroup, metricLabel, resources.MetricDelete, true, true)
+
+	return err
+}
+
+func (k *K8sWatcher) updateCIDRGroupRefPolicies(
+	cidrGroup string,
+	cidrGroupCache map[string]*cilium_v2_alpha1.CiliumCIDRGroup,
+	cnpCache map[resource.Key]*types.SlimCNP,
+	cs client.Clientset,
+) error {
+	var errs []error
+	for key, cnp := range cnpCache {
```

Comment:
Could there be any performance concerns with a cluster with a lot of CNP and Refs. We're also duplicating a lot of the cnp watch.Store with this.

Would it make any sense to store a ref index onto CNPs?

Reply:
Or perhaps we could index on ref somehow, similar to CEP/CES?

Reply:
^ Might be a good case for some sort of Go benchmark to get a sense for the performance of it. We do have the metrics at least to be able to measure the impact.

Reply:
Yeah don't want to encourage premature optimization, but it might be worth due to it being tied up with the CNP update path now.

Reply:
We discussed this item a bit more:

We determined policy implementation delay is more concerning in terms of impact of this code instead of the CNP translation time code. We don't expect the CNP translation time to have a significant impact here. Given the state of the controlplane tests and that fact that doesn't yet support endpoint regeneration due to lack of datapath mocking, we can't yet devise controlplane test to stress this code without doing it manually. So the idea would be to create the resources manually and observe policy implementation delay metric. 

Memory is not the concern because the policies and the CIDRs within the policy are at maximum ~1KB and 1000 of these policies are only ~1MB. The concern would be more CPU.

The impact to user if we have a performance issue here is increased delay for policy recalculation.

I added this as something to followup on in https://github.com/cilium/cilium/issues/24801.